src/snowflake/cli/api/entities/application_package_entity.py

```diff
@@ -161,6 +163,36 @@ def action_drop(
             force_drop=force_drop,
         )
 
+    def action_validate(self, ctx: ActionContext):
+        model = self._entity_model
+        package_name = model.fqn.identifier
+        stage_fqn = f"{package_name}.{model.stage}"
+        if model.meta and model.meta.role:
+            package_role = model.meta.role
+        else:
+            package_role = ctx.default_role
+
+        def deploy_to_scratch_stage_fn():
+            self.action_deploy(
+                ctx=ctx,
+                prune=True,
+                recursive=True,
+                paths=[],
+                validate=False,
+                stage_fqn=model.scratch_stage,
+            )
+
+        self.validate_setup_script(
+            console=ctx.console,
+            package_name=package_name,
+            package_role=package_role,
+            stage_fqn=stage_fqn,
+            use_scratch_stage=True,
+            scratch_stage_fqn=model.scratch_stage,
```

Comment:
why do we have to pass both stage_fqn and scratch_stage_fqn. Don't we only use one of the 2?

Reply:
Yes, I considered simplifying it -- that's how it was implemented in `NativeAppManager`. But I prefer keeping the signatures and logic the same to avoid making multiple changes in this refactor.